File: lib/hypercube.ts
// user of field.
import {
    Map,
} from 'immutable';
import Field, {
    State,
    ICoeff,
    makeState,
    getDir,
    getPos,
} from './field';
import {
    Complex,
    cadd,
    csmul,
    czero,
    cone,
} from './complex';
import {
    positionalDistribution,
    showPositionalDistribution,
} from './util';

declare var require : (path:string)=>any;

const fs = require('fs');
const path = require('path');
const appRootPath = require('app-root-path');

// 結果を出力するところ
const OUTPATH = path.join(appRootPath.toString(), 'data.txt');

// 次元
const MAXN = 8;

// 結果出力先ファイルをオープン
const filestream = fs.createWriteStream(OUTPATH, {
    flags: 'w',
    defaultEncoding: 'utf8',
});

// 各次元ごとに計測する
for(let N = 1; N <= MAXN; N++){
    console.log(`===== ${N} =====`);
    // サンプル数
    const iter = Math.max(16, 2*Math.floor(N**2));

    // Fieldを初期化
    const director = (d: Map<State, Complex>)=>{
        const result = Map<State, Complex>().asMutable();
        // 方向
        for(let a=0; a<N; a++){
            // 位置
            for(let v=0, n=2**N; v<n; v++){
                // Sを適用する前の位置
                const vb = v ^ (1 << a);
                // vbの確率はGrover's diffusion operatorをかけて求める
                let s = czero;
                for(let b=0; b<N; b++){
                    const dd = a===b ? 2/N-1 : 2/N;
                    s = cadd(s, csmul(dd, d.get(makeState(b, vb), czero)));
                }
                result.set(makeState(a, v), s);
            }
        }
        return result.asImmutable();
    };
    const f = new Field(director);

    // 初期状態
    const coeff = Map<State, Complex>().set(makeState(0, 0), cone) as ICoeff;

    let cnt = 0;
    let steps = 0;
    for(let i=0; i<iter; i++){
        console.log(i);
        // absorbing vertexはランダムにset
        const m = Math.floor(Math.random() * (2**N));
        // antipodalな位置
        // const m = (2**N) - 1;

        f.init(coeff);
        // step数
        const s = f.test(m, 180);

        if(s >= 0){
            cnt++;
            steps += s;
        }
    }
    // absorbing vertexに到達する確率
    const p = cnt / iter;
    // absorbing time
    const time = steps / cnt / p;

    filestream.write(`${N} ${cnt} ${p} ${time}\n`);
    console.log(N, p, time);
}

filestream.end();
console.log('end', new Date());

File: c/hypercube.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <complex.h>
#include "MT.h"

#define MAXN (10)

#define WALKITER (240)

void simulate(int n);
int walktest(double complex *S1, double complex *S2, double complex *op, int n, int vn, int absorb);
void matvecmul(double complex *out, double complex *A, double complex *v, int m, int n);
void normalize(double complex *v, int vn);

int main(){
	int n;
	init_genrand(100);
	for(n=1; n<=MAXN; n++){
		fprintf(stderr, "===== %d =====\n", n);
		simulate(n);
	}
	return 0;
}


void simulate(int n){
	int iter, i, j, k;
	int vn, steps, stepsum, cnt;
	double complex *S1, *S2, *op;
	double p, time;
	iter = (int)floor(1.8*(double)(n*n));

	//stateベクトルの確保(N * 2^N)
	vn = (1<<n)*n;

	S1 = (double complex*)malloc(vn *(sizeof (double complex)));
	S2 = (double complex*)malloc(vn *(sizeof (double complex)));
	//オペレーターの確保
	op = (double complex*)malloc(vn*vn *(sizeof (double complex)));

	/* オペレーターを計算 */
	memset(op, 0, vn*vn*sizeof(double complex));
	for(i=0; i<vn; i++){
		for(j=0; j<vn; j++){
			for(k=0; k<n; k++){
				op[(k*(1<<n)+i)*vn + (k*(1<<n)+j)] = i==j ? 2.0/(double)n - 1.0 : 2.0/(double)n;
			}
		}
	}

	stepsum = 0;
	cnt = 0;
	for(i=0; i<iter; i++){
		fprintf(stderr, "%d\n", i);
		/* 初期状態を作る */
		S1[0] = (double complex)1.0;
		for(j=1; j<vn; j++){
			/* 初期状態: |1, 0....0> */
			S1[j] = (double complex)0.0;
		}

		/* walk testする */
		steps = walktest(S1, S2, op, n, vn, (1<<n)-1);
		printf("%d\n", steps);
		if(steps >= 0){
			cnt++;
			stepsum += steps;
		}
	}
	/* absorbing vertexに到達する確率 */
	p = cnt / (double)iter;
	time = (double)stepsum / (double)cnt / p;
	printf("%d %d %f %f\n", n, cnt, p, time);

	free(S1);
	free(S2);
	free(op);
}

int walktest(double complex *S1, double complex *S2, double complex *op, int n, int vn, int absorb){
	/* walkしてtestする */
	int i, j;
	double complex *tmp;
	double sum, a, r;
	for(i=0; i<WALKITER; i++){
		/* 1回walkする */
		matvecmul(S2, op, S1, vn, vn);
		/* S1とS2を交代 */
		tmp = S2;
		S2 = S1;
		S1 = tmp;
		/* S1をnormalizeする */
		normalize(S1, vn);
		/* 計測してabsorbing vertexにいるか確認 */
		sum = 0.0;
		for(j=0; j<n; j++){
			/* 各方向のabsorbing vertex */
			a = cabs(S1[j*(1<<n)+absorb]);
			sum += a*a;
		}
		/* 確率 */
		r = genrand_real2();
		if(r < sum){
			// 到達！
			return i;
		}
		/* 到達しなかったから消す */
		for(j=0; j<n; j++){
			S1[j*(1<<n)+absorb] = 0.0;
		}
		normalize(S1, vn);
	}
	return -1;
}

/* m*n行列と長さnのベクトルをかける */
void matvecmul(double complex *out, double complex *A, double complex *v, int m, int n){
	int i, j;
	for(i=0; i<m; i++){
		out[i] = (double complex)0.0;
		for(j=0; j<n; j++){
			out[i] += A[i*n+j]*v[j];
		}
	}
}
/* 長さnのベクトルの長さを1に直す */
void normalize(double complex *v, int n){
	int i;
	double sum = 0, a;
	for(i=0; i<n; i++){
		a = cabs(v[i]);
		sum += a*a;
	}
	if(fabs(sum - 1.0) >= 0.000001){
		a = 1.0 / sum;
		for(i=0; i<n; i++){
			v[i] *= a;
		}
	}
}
